### PR TITLE
Parametrize overprovisioning and upgrade mode

### PR DESCRIPTION
**Pull request by r7vme**
- parametrize upgrade mode and overprovisioning ([we can enable "Automatic" mode actually](https://github.com/giantswarm/azure-operator/issues/238#issuecomment-397593016))
- disable overprovisioning for masters

NOTE:
Potentially we want to replace "update" with "reimage" (introduced here https://github.com/giantswarm/azure-operator/commit/426fbb4e469319a430d663a545738b3334a5c8e1 ) as actual upgrade for us done, when vm "reimaged" and not when "upgdaded"

### CODE REVIEW

File: service/controller/v2/arm_templates/vmss.json [hunk 1]
```diff
@@ -29,7 +29,7 @@
             "type": "String"
         },
         "vmssSshPublicKey": {
-            "type": "String"
+            "type": "string"
```
Comment (r7vme) [inline service/controller/v2/arm_templates/vmss.json]: boyscouting, use lower case everywhere

File: service/controller/v2/arm_templates/main.json [hunk 2]
```diff
@@ -362,12 +362,14 @@
           "vmssOsImageOffer": {"value": "[parameters('masterNodes')[0].osImage.offer]"},
           "vmssOsImageSKU": {"value": "[parameters('masterNodes')[0].osImage.sku]"},
           "vmssOsImageVersion": {"value": "[parameters('masterNodes')[0].osImage.version]"},
+          "vmssOverprovision": {"value": "false"},
```
Comment (r7vme) [inline service/controller/v2/arm_templates/main.json]: we don't need overprovision for masters as by default it creates two vms and both can be up at the same time, messing up load balancers.

**Comment by xh3b4sd** [comment]
I do not get the full picture here. What does disabling overprovisioning for masters and enabling it for workers do for us now? Is that just an intermediate step or do we fix some problem now?

**Comment by r7vme** [comment]
> I do not get the full picture here. What does disabling overprovisioning for masters and enabling it for workers do for us now? 

Overprovisioning is enabled by default. We just disabling overprovisioning for masters. Imagine that two masters coming up and registered in load balancers with completely different etcd disks. We want to prevent such situations. This also makes master ID predictable, it will always be `0`. Right now we have a race between `0` and `1` and one that will faster report ready status (from waagent) wins. But as i said it's not we actually need.

Keeping overprovisioning for workers makes sense, because it's makes vmss creation with large number of VMs more reliable. Moreover if VM will be able to register in k8s at some point it will be removed by Kubernetes as soon as it will be removed by vmss (when 10/10 vms will be ready, others are removed).

From official docs
> You can overprovision scale set VMs for increased reliability and quicker deployment times. You cannot overprovision individual VMs unless you write custom code to perform this action.

**Comment by r7vme** [comment]
> I do not get the full picture here. 

1. overprovisioning (see comment above) this is the final change - we want to disable over provisioning for master. This is done here.

2. upgrade mode. 

Nothing changed in actual behavious, we still on "Manual" mode, but this change makes upgrade mode configurable. Finally i expect that we will use "Automatic" mode. Is it clear why we want to use "Automatic" mode?

**Comment by xh3b4sd** [comment]
> Is it clear why we want to use "Automatic" mode?

Not really. AFAIK we wanted to use "Manual" to have full control of the update process because we need to hook into it in order to drain nodes. I am not sure right now which implications "Automatic" has.

**Comment by r7vme** [comment]
> Not really. AFAIK we wanted to use "Manual" to have full control of the update process because we need to hook into it in order to drain nodes. I am not sure right now which implications "Automatic" has.

Actually good point. We need to do few more tests before enabling "Automatic". I tested if custom_data changed nothing is happens. Just latest metadata applied to VM (so on next reimage new custom_data) will be used. We also need to test on what happens if vm size or CoreOS image changed.

File: service/controller/v2/arm_templates/main.json [hunk 3]
```diff
@@ -397,9 +399,11 @@
           "vmssOsImageOffer": {"value": "[parameters('workerNodes')[0].osImage.offer]"},
           "vmssOsImageSKU": {"value": "[parameters('workerNodes')[0].osImage.sku]"},
           "vmssOsImageVersion": {"value": "[parameters('workerNodes')[0].osImage.version]"},
+          "vmssOverprovision": {"value": "true"},
           "vmssVmCustomData": {"value": "[parameters('workerCloudConfigData')]"},
           "vmssLbBackendPools": {"value": []},
-          "vmssVnetSubnetId": {"value": "[reference('virtual_network_setup').outputs.workerSubnetID.value]"}
+          "vmssVnetSubnetId": {"value": "[reference('virtual_network_setup').outputs.workerSubnetID.value]"},
+          "vmssUpgradePolicy": {"value": "Manual"}
```
Comment (kopiczko) [inline service/controller/v2/arm_templates/main.json]: Why this is parametrized?

Comment (r7vme) [reply service/controller/v2/arm_templates/main.json]: It's just a place holder if we will want to outsource upgrades to Azure. From latest discussions seems that we will do both "upgrade" and "reimage" in operator. So this is not really needed now.

**Pull request merged**